File: src/main/java/org/jukeboxmc/world/chunk/manager/ChunkManager.java
package org.jukeboxmc.world.chunk.manager;

import com.google.common.collect.ImmutableSet;
import com.spotify.futures.CompletableFutures;
import it.unimi.dsi.fastutil.longs.Long2LongMap;
import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.objects.ObjectIterator;
import lombok.ToString;
import lombok.extern.log4j.Log4j2;
import org.jukeboxmc.Server;
import org.jukeboxmc.event.world.ChunkLoadEvent;
import org.jukeboxmc.event.world.ChunkUnloadEvent;
import org.jukeboxmc.util.Utils;
import org.jukeboxmc.world.Dimension;
import org.jukeboxmc.world.World;
import org.jukeboxmc.world.chunk.Chunk;

import javax.annotation.ParametersAreNonnullByDefault;
import java.rmi.ServerError;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;

@Log4j2
@ParametersAreNonnullByDefault
public final class ChunkManager {
    private static final CompletableFuture<Void> COMPLETED_VOID_FUTURE = CompletableFuture.completedFuture( null );
    private static final AtomicIntegerFieldUpdater<LoadingChunk> GENERATION_RUNNING_UPDATER = AtomicIntegerFieldUpdater.newUpdater( LoadingChunk.class, "generationRunning" );
    private static final AtomicIntegerFieldUpdater<LoadingChunk> POPULATION_RUNNING_UPDATER = AtomicIntegerFieldUpdater.newUpdater( LoadingChunk.class, "populationRunning" );
    private static final AtomicIntegerFieldUpdater<LoadingChunk> FINISH_RUNNING_UPDATER = AtomicIntegerFieldUpdater.newUpdater( LoadingChunk.class, "finishRunning" );

    private final World world;
    private final Dimension dimension;
    private final Long2ObjectMap<LoadingChunk> chunks = new Long2ObjectOpenHashMap<>();
    private final Long2LongMap chunkLoadedTimes = new Long2LongOpenHashMap();
    private final Long2LongMap chunkLastAccessTimes = new Long2LongOpenHashMap();
    private final Executor executor;

    public ChunkManager( World world, Dimension dimension ) {
        this.world = world;
        this.dimension = dimension;
        this.executor = this.world.getServer().getScheduler().getChunkExecutor();
    }

    public synchronized Set<Chunk> getLoadedChunks() {
        ImmutableSet.Builder<Chunk> chunks = ImmutableSet.builder();
        for ( LoadingChunk loadingChunk : this.chunks.values() ) {
            Chunk chunk = loadingChunk.getChunk();
            if ( chunk != null ) {
                chunks.add( chunk );
            }
        }
        return chunks.build();
    }

    public synchronized int getLoadedCount() {
        return this.chunks.size();
    }

    public synchronized Chunk getLoadedChunk( long key ) {
        LoadingChunk chunk = this.chunks.get( key );
        return chunk == null ? null : chunk.getChunk();
    }

    public synchronized Chunk getLoadedChunk( int x, int z ) {
        return getLoadedChunk( Utils.toLong( x, z ) );
    }

    public Chunk getChunk( int x, int z ) {
        try {
            Chunk chunk = this.getLoadedChunk( x, z );
            if ( chunk == null ) {
                chunk = this.getChunkFuture( x, z ).join();
            }
            return chunk;
        } catch ( Exception e ) {
            e.printStackTrace();
        }
        return null;
    }

    public CompletableFuture<Chunk> getChunkFuture( int x, int z ) {
        return this.getChunkFuture( x, z, true, true, true );
    }

    private synchronized CompletableFuture<Chunk> getChunkFuture( int chunkX, int chunkZ, boolean generate, boolean populate, boolean finish ) {
        final long chunkKey = Utils.toLong( chunkX, chunkZ );
        this.chunkLastAccessTimes.put( chunkKey, System.currentTimeMillis() );
        LoadingChunk chunk = this.chunks.computeIfAbsent( chunkKey, key -> new LoadingChunk( this.dimension, key, true ) );

        if ( finish ) {
            chunk.finish();
        } else if ( populate ) {
            chunk.populate();
        } else if ( generate ) {
            chunk.generate();
        }

        Server.getInstance().getPluginManager().callEvent( new ChunkLoadEvent( this.world, chunk.getChunk() ) );

        return chunk.getFuture();
    }

    public synchronized boolean isChunkLoaded( long hash ) {
        LoadingChunk chunk = this.chunks.get( hash );
        return chunk != null && chunk.getChunk() != null;
    }

    public synchronized boolean isChunkLoaded( int x, int z ) {
        return this.isChunkLoaded( Utils.toLong( x, z ) );
    }

    public synchronized boolean unloadChunk( long hash ) {
        return this.unloadChunk( hash, true, true );
    }

    public boolean unloadChunk( Chunk chunk ) {
        return this.unloadChunk( chunk, true );
    }

    public boolean unloadChunk( Chunk chunk, boolean save ) {
        return this.unloadChunk( chunk, save, true );
    }

    public boolean unloadChunk( Chunk chunk, boolean save, boolean safe ) {
        return this.unloadChunk( Utils.toLong( chunk.getX(), chunk.getZ() ), save, safe );
    }

    public boolean unloadChunk( long chunkKey, boolean save, boolean safe ) {
        Chunk loadedChunk = this.getLoadedChunk( chunkKey );
        if ( loadedChunk == null ) return false;
        ChunkUnloadEvent chunkUnloadEvent = new ChunkUnloadEvent( this.world, loadedChunk, save );
        Server.getInstance().getPluginManager().callEvent( chunkUnloadEvent );
        if ( chunkUnloadEvent.isCancelled() ) return false;
        boolean result = unloadChunk0( chunkKey, chunkUnloadEvent.isSaveChunk(), safe );
        if ( result ) {
            this.chunks.remove( chunkKey );
        }
        return result;
    }

    private synchronized boolean unloadChunk0( long chunkKey, boolean save, boolean safe ) {
        LoadingChunk loadingChunk = this.chunks.get( chunkKey );
        if ( loadingChunk == null ) {
            return false;
        }
        Chunk chunk = loadingChunk.getChunk();
        if ( chunk == null ) {
            return false;
        }
        if ( !chunk.getLoaders().isEmpty() ) {
            return false;
        }

        if ( save ) {
            this.saveChunk( chunk );
        }

        if ( safe && !this.world.getChunkPlayers( chunk.getX(), chunk.getZ(), chunk.getDimension() ).isEmpty() ) {
            return false;
        }

        this.chunkLastAccessTimes.remove( chunkKey );
        this.chunkLoadedTimes.remove( chunkKey );

        return true;
    }

    public synchronized CompletableFuture<Void> saveChunks() {
        List<CompletableFuture<?>> futures = new ArrayList<>();
        for ( LoadingChunk loadingChunk : this.chunks.values() ) {
            Chunk chunk = loadingChunk.getChunk();
            if ( chunk != null ) {
                futures.add( saveChunk( chunk ) );
            }
        }

        return CompletableFuture.allOf( futures.toArray( new CompletableFuture[0] ) );
    }

    public CompletableFuture<Void> saveChunk( Chunk chunk ) {
        if ( chunk.isDirty() ) {
            return this.world.saveChunk( chunk ).exceptionally( throwable -> {
                log.warn( "Unable to save chunk", throwable );
                return null;
            } );
        }
        return COMPLETED_VOID_FUTURE;
    }

    public synchronized void tick() {
        if ( this.chunks.isEmpty() ) {
            return;
        }

        long time = System.currentTimeMillis();

        // Spawn chunk
        final int spawnX = this.world.getSpawnLocation().getChunkX();
        final int spawnZ = this.world.getSpawnLocation().getChunkZ();
        final int spawnRadius = Server.getInstance().getViewDistance();

        // Do chunk garbage collection
        ObjectIterator<Long2ObjectMap.Entry<LoadingChunk>> iterator = this.chunks.long2ObjectEntrySet().iterator();
        while ( iterator.hasNext() ) {
            Long2ObjectMap.Entry<LoadingChunk> entry = iterator.next();
            long chunkKey = entry.getLongKey();
            LoadingChunk loadingChunk = entry.getValue();
            Chunk chunk = loadingChunk.getChunk();
            if ( chunk == null ) {
                continue; // Chunk hasn't loaded
            }

            if ( ( Math.abs( chunk.getX() - spawnX ) <= spawnRadius && Math.abs( chunk.getZ() - spawnZ ) <= spawnRadius ) ||
                    !chunk.getLoaders().isEmpty() ) {
                continue; // Spawn protection or is loaded
            }

            long loadedTime = this.chunkLoadedTimes.get( chunkKey );
            if ( ( time - loadedTime ) <= TimeUnit.SECONDS.toMillis( 30 ) ) {
                continue;
            }

            long lastAccessTime = this.chunkLastAccessTimes.get( chunkKey );
            if ( ( time - lastAccessTime ) <= TimeUnit.SECONDS.toMillis( 120 ) ) {
                continue;
            }

            ChunkUnloadEvent chunkUnloadEvent = new ChunkUnloadEvent( this.world, chunk, true );
            Server.getInstance().getPluginManager().callEvent( chunkUnloadEvent );

            if ( chunkUnloadEvent.isCancelled() ) {
                return;
            }

            if ( this.unloadChunk0( chunkKey, chunkUnloadEvent.isSaveChunk(), true ) ) {
                iterator.remove();
            }
        }
    }

    @ToString
    private class LoadingChunk {

        private final int x;
        private final int z;
        private CompletableFuture<Chunk> future;
        volatile int generationRunning;
        volatile int populationRunning;
        volatile int finishRunning;
        private Chunk chunk;

        public LoadingChunk( Dimension dimension, long key, boolean load ) {
            this.x = Utils.fromHashX( key );
            this.z = Utils.fromHashZ( key );

            if ( load ) {
                this.future = ChunkManager.this.world.readChunk( new Chunk( ChunkManager.this.world, dimension, this.x, this.z ) )
                        .thenApply( chunk -> Objects.requireNonNullElseGet( chunk, () -> new Chunk( ChunkManager.this.world, dimension, this.x, this.z ) ) );
                this.future.whenComplete( ( chunk, throwable ) -> {
                    if ( throwable != null ) {
                        Server.getInstance().getLogger().error( "Unable to load chunk " + x + ":" + z, throwable );

                        synchronized (ChunkManager.this) {
                            ChunkManager.this.chunks.remove( key );
                        }
                    } else {
                        long currentTime = System.currentTimeMillis();
                        synchronized (ChunkManager.this) {
                            ChunkManager.this.chunkLoadedTimes.put( key, currentTime );
                        }
                    }
                } );
            } else {
                this.future = CompletableFuture.completedFuture( new Chunk( ChunkManager.this.world, dimension, this.x, this.z ) );
            }
            this.future.whenComplete( ( chunk, throwable ) -> this.chunk = chunk );
        }

        public CompletableFuture<Chunk> getFuture() {
            return future;
        }

        private Chunk getChunk() {
            if ( this.chunk != null && this.chunk.isGenerated() && this.chunk.isPopulated() && this.chunk.isFinished() ) {
                return this.chunk;
            }
            return null;
        }

        private void generate() {
            if ( ( this.chunk == null || !this.chunk.isGenerated() ) && GENERATION_RUNNING_UPDATER.compareAndSet( this, 0, 1 ) ) {
                future = future.thenApplyAsync( GenerationTask.INSTANCE, ChunkManager.this.executor );
                future.thenRun( () -> GENERATION_RUNNING_UPDATER.compareAndSet( this, 1, 0 ) );
            }
        }

        private void populate() {
            this.generate();
            if ( ( this.chunk == null || !this.chunk.isPopulated() ) && POPULATION_RUNNING_UPDATER.compareAndSet( this, 0, 1 ) ) {
                // Load and generate chunks around the chunk to be populated.
                List<CompletableFuture<Chunk>> chunksToLoad = new ArrayList<>( 8 );
                for ( int z = this.z - 1, maxZ = this.z + 1; z <= maxZ; z++ ) {
                    for ( int x = this.x - 1, maxX = this.x + 1; x <= maxX; x++ ) {
                        if ( x == this.x && z == this.z ) continue;
                        chunksToLoad.add( ChunkManager.this.getChunkFuture( x, z, true, false, false ) );
                    }
                }
                CompletableFuture<List<Chunk>> aroundFuture = CompletableFutures.allAsList( chunksToLoad );

                future = future.thenCombineAsync( aroundFuture, PopulationTask.INSTANCE, ChunkManager.this.executor );
                future.thenRun( () -> POPULATION_RUNNING_UPDATER.compareAndSet( this, 1, 0 ) );
            }
        }

        private void finish() {
            this.populate();
            if ( ( this.chunk == null || !this.chunk.isFinished() ) && FINISH_RUNNING_UPDATER.compareAndSet( this, 0, 1 ) ) {
                List<CompletableFuture<Chunk>> chunksToLoad = new ArrayList<>( 8 );
                for ( int z = this.z - 1, maxZ = this.z + 1; z <= maxZ; z++ ) {
                    for ( int x = this.x - 1, maxX = this.x + 1; x <= maxX; x++ ) {
                        if ( x == this.x && z == this.z ) continue;
                        chunksToLoad.add( ChunkManager.this.getChunkFuture( x, z, true, true, false ) );
                    }
                }
                CompletableFuture<List<Chunk>> aroundFuture = CompletableFutures.allAsList( chunksToLoad );

                future = future.thenCombineAsync( aroundFuture, FinishingTask.INSTANCE, ChunkManager.this.executor );
                future.thenRun( () -> FINISH_RUNNING_UPDATER.compareAndSet( this, 1, 0 ) );
            }
        }

        private void clear() {
            this.future = future.thenApply( chunk -> {
                GENERATION_RUNNING_UPDATER.set( this, 0 );
                POPULATION_RUNNING_UPDATER.set( this, 0 );
                FINISH_RUNNING_UPDATER.set( this, 0 );
                return chunk;
            } );
        }
    }

    public Dimension getDimension() {
        return this.dimension;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/world/chunk/manager/ChunkManager.kt
package org.jukeboxmc.server.world.chunk.manager

import com.google.common.collect.ImmutableSet
import com.spotify.futures.CompletableFutures
import it.unimi.dsi.fastutil.longs.*
import org.jukeboxmc.api.event.world.ChunkLoadEvent
import org.jukeboxmc.api.event.world.ChunkUnloadEvent
import org.jukeboxmc.api.world.Dimension
import org.jukeboxmc.api.world.chunk.Chunk
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.scheduler.JukeboxScheduler
import org.jukeboxmc.server.util.Utils
import org.jukeboxmc.server.world.JukeboxWorld
import org.jukeboxmc.server.world.chunk.JukeboxChunk
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater
import kotlin.concurrent.Volatile
import kotlin.math.abs

class ChunkManager(
    private val world: JukeboxWorld,
    private val dimension: Dimension
) {

    private val chunks: Long2ObjectMap<LoadingChunk> = Long2ObjectOpenHashMap()
    private val chunkLoadedTimes: Long2LongMap = Long2LongOpenHashMap()
    private val chunkLastAccessTimes: Long2LongMap = Long2LongOpenHashMap()
    private val executor: Executor = JukeboxScheduler.getInstance().getChunkExecutor()

    companion object {
        private val COMPLETED_VOID_FUTURE = CompletableFuture.completedFuture<Void?>(null)
    }

    @Synchronized
    fun getLoadedChunks(): Set<Chunk> {
        val chunks = ImmutableSet.builder<Chunk>()
        for (loadingChunk in this.chunks.values) {
            val chunk = loadingChunk.getChunk()
            if (chunk != null) {
                chunks.add(chunk)
            }
        }
        return chunks.build()
    }

    @Synchronized
    fun getLoadedCount(): Int {
        return chunks.size
    }

    @Synchronized
    fun getLoadedChunk(key: Long): JukeboxChunk? {
        val chunk = chunks[key]
        return chunk?.getChunk()
    }

    @Synchronized
    fun getLoadedChunk(x: Int, z: Int): JukeboxChunk? {
        return getLoadedChunk(Utils.toLong(x, z))
    }

    fun getChunk(x: Int, z: Int): Chunk? {
        try {
            var chunk = this.getLoadedChunk(x, z)
            if (chunk == null) {
                chunk = this.getChunkFuture(x, z).join()
            }
            return chunk
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    @Synchronized
    fun getChunkFuture(
        chunkX: Int,
        chunkZ: Int,
        generate: Boolean = true,
        populate: Boolean = true,
        finish: Boolean = true
    ): CompletableFuture<JukeboxChunk> {
        val chunkKey = Utils.toLong(chunkX, chunkZ)
        chunkLastAccessTimes.put(chunkKey, System.currentTimeMillis())
        val chunk = chunks.computeIfAbsent(chunkKey,
            Long2ObjectFunction { key: Long ->
                LoadingChunk(
                    this,
                    dimension,
                    key,
                    true
                )
            })
        if (finish) {
            chunk.finish()
        } else if (populate) {
            chunk.populate()
        } else if (generate) {
            chunk.generate()
        }
        JukeboxServer.getInstance().getPluginManager().callEvent(ChunkLoadEvent(world, chunk.getChunk()))
        return chunk.getFuture()
    }

    @Synchronized
    fun isChunkLoaded(hash: Long): Boolean {
        val chunk = chunks[hash]
        return chunk?.getChunk() != null
    }

    @Synchronized
    fun isChunkLoaded(x: Int, z: Int): Boolean {
        return this.isChunkLoaded(Utils.toLong(x, z))
    }

    @Synchronized
    fun unloadChunk(hash: Long): Boolean {
        return this.unloadChunk(hash, save = true, safe = true)
    }

    fun unloadChunk(chunk: Chunk): Boolean {
        return this.unloadChunk(chunk, true)
    }

    fun unloadChunk(chunk: Chunk, save: Boolean): Boolean {
        return this.unloadChunk(chunk, save, true)
    }

    fun unloadChunk(chunk: Chunk, save: Boolean, safe: Boolean): Boolean {
        return this.unloadChunk(Utils.toLong(chunk.getX(), chunk.getZ()), save, safe)
    }

    fun unloadChunk(chunkKey: Long, save: Boolean, safe: Boolean): Boolean {
        val loadedChunk = this.getLoadedChunk(chunkKey) ?: return false
        val chunkUnloadEvent = ChunkUnloadEvent(world, loadedChunk, save)
        JukeboxServer.getInstance().getPluginManager().callEvent(chunkUnloadEvent)
        if (chunkUnloadEvent.isCancelled()) return false
        val result = unloadChunk0(chunkKey, chunkUnloadEvent.saveChunk(), safe)
        if (result) {
            this.chunks.remove(chunkKey)
        }
        return result
    }

    @Synchronized
    private fun unloadChunk0(chunkKey: Long, save: Boolean, safe: Boolean): Boolean {
        val loadingChunk = chunks[chunkKey] ?: return false
        val chunk = loadingChunk.getChunk() ?: return false
        if (chunk.getLoaders().isNotEmpty()) {
            return false
        }
        if (save) {
            this.saveChunk(chunk)
        }
        if (safe && world.getChunkPlayers(chunk.getX(), chunk.getZ(), chunk.getDimension()).isNotEmpty()) {
            return false
        }
        chunkLastAccessTimes.remove(chunkKey)
        chunkLoadedTimes.remove(chunkKey)
        return true
    }

    @Synchronized
    fun saveChunks(): CompletableFuture<Void?>? {
        val futures: MutableList<CompletableFuture<*>> = ArrayList()
        for (loadingChunk in chunks.values) {
            val chunk: JukeboxChunk? = loadingChunk.getChunk()
            if (chunk != null) {
                futures.add(saveChunk(chunk))
            }
        }
        return CompletableFuture.allOf(*futures.toTypedArray<CompletableFuture<*>>())
    }

    fun saveChunk(chunk: JukeboxChunk): CompletableFuture<Void?> {
        return if (chunk.isDirty()) {
            world.getLevelDBStorage().saveChunk(chunk).exceptionally { _ ->
                JukeboxServer.getInstance().getLogger().info("Unable to save chunk")
                null
            }
        } else COMPLETED_VOID_FUTURE
    }

    @Synchronized
    fun tick() {
        if (chunks.isEmpty()) {
            return
        }
        val time = System.currentTimeMillis()

        // Spawn chunk
        val spawnX: Int = world.getSpawnLocation(this.dimension).getChunkX()
        val spawnZ: Int = world.getSpawnLocation(this.dimension).getChunkZ()
        val spawnRadius: Int = 4 //TODO UPDATE
        //val spawnRadius: Int = JukeboxServer.getInstance().getViewDistance()

        // Do chunk garbage collection
        val iterator = chunks.long2ObjectEntrySet().iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            val chunkKey = entry.longKey
            val loadingChunk = entry.value
            val chunk = loadingChunk.getChunk()
                ?: continue  // Chunk hasn't loaded
            if (abs(chunk.getX() - spawnX) <= spawnRadius && abs(chunk.getZ() - spawnZ) <= spawnRadius || chunk.getLoaders().isNotEmpty()) {
                continue  // Spawn protection or is loaded
            }
            val loadedTime = chunkLoadedTimes[chunkKey]
            if (time - loadedTime <= TimeUnit.SECONDS.toMillis(30)) {
                continue
            }
            val lastAccessTime = chunkLastAccessTimes[chunkKey]
            if (time - lastAccessTime <= TimeUnit.SECONDS.toMillis(120)) {
                continue
            }
            val chunkUnloadEvent = ChunkUnloadEvent(world, chunk, true)
            JukeboxServer.getInstance().getPluginManager().callEvent(chunkUnloadEvent)
            if (chunkUnloadEvent.isCancelled()) {
                return
            }
            if (unloadChunk0(chunkKey, chunkUnloadEvent.saveChunk(), true)) {
                iterator.remove()
            }
        }
    }

    class LoadingChunk(
        private val chunkManager: ChunkManager,
        private val dimension: Dimension,
        private val key: Long,
        private val load: Boolean
    ) {
        private val x: Int = Utils.fromHashX(key)
        private val z: Int = Utils.fromHashZ(key)
        private var future: CompletableFuture<JukeboxChunk>
        private var chunk: JukeboxChunk? = null

        @Volatile
        var generationRunning = 0

        @Volatile
        var populationRunning = 0

        @Volatile
        var finishRunning = 0

        companion object {
            private val GENERATION_RUNNING_UPDATER: AtomicIntegerFieldUpdater<LoadingChunk> =
                AtomicIntegerFieldUpdater.newUpdater(LoadingChunk::class.java, "generationRunning")
            private val POPULATION_RUNNING_UPDATER: AtomicIntegerFieldUpdater<LoadingChunk> =
                AtomicIntegerFieldUpdater.newUpdater(LoadingChunk::class.java, "populationRunning")
            private val FINISH_RUNNING_UPDATER: AtomicIntegerFieldUpdater<LoadingChunk> =
                AtomicIntegerFieldUpdater.newUpdater(LoadingChunk::class.java, "finishRunning")
        }

        init {
            if (this.load) {
                this.future =
                    this.chunkManager.world.getLevelDBStorage()
                        .readChunk(JukeboxChunk(this.chunkManager.world, this.dimension, this.x, this.z))
                        .thenApply { chunk ->
                            chunk ?: JukeboxChunk(this.chunkManager.world, this.dimension, this.x, this.z)
                        }
                this.future.whenComplete { _, throwable ->
                    if (throwable != null) {
                        throwable.printStackTrace()
                        JukeboxServer.getInstance().getLogger().error("Unable to load chunk $x:$z")

                        synchronized(ChunkManager) {
                            this.chunkManager.chunks.remove(key)
                        }
                    } else {
                        val currentTime = System.currentTimeMillis()
                        synchronized(ChunkManager) {
                            this.chunkManager.chunkLoadedTimes[key] = currentTime
                        }
                    }
                }
            } else {
                this.future = CompletableFuture.completedFuture(
                    JukeboxChunk(
                        this.chunkManager.world,
                        this.dimension,
                        this.x,
                        this.z
                    )
                )
            }
            this.future.whenComplete { loadedChunk, th ->
                th?.printStackTrace()
                this.chunk = loadedChunk
            }
        }

        fun getFuture(): CompletableFuture<JukeboxChunk> {
            return this.future
        }

        fun getChunk(): JukeboxChunk? {
            if (this.chunk != null && this.chunk!!.isGenerated() && this.chunk!!.isPopulated() && this.chunk!!.isFinished()) {
                return this.chunk!!
            }
            return null
        }

        fun generate() {
            if ((this.chunk == null || !this.chunk!!.isGenerated()) && GENERATION_RUNNING_UPDATER.compareAndSet(
                    this,
                    0,
                    1
                )
            ) {
                this.future = this.future.thenApplyAsync(GenerationTask.INSTANCE, chunkManager.executor)
                this.future.thenRun { GENERATION_RUNNING_UPDATER.compareAndSet(this, 1, 0) }
            }
        }

        fun populate() {
            generate()
            if ((this.chunk == null || !this.chunk!!.isPopulated()) && POPULATION_RUNNING_UPDATER.compareAndSet(
                    this,
                    0,
                    1
                )
            ) {
                val chunksToLoad = ArrayList<CompletableFuture<JukeboxChunk>>(8)
                for (z in this.z - 1..this.z + 1) {
                    for (x in this.x - 1..this.x + 1) {
                        if (x == this.x && z == this.z) continue
                        chunksToLoad.add(this.chunkManager.getChunkFuture(x, z,
                            generate = true,
                            populate = false,
                            finish = false
                        ))
                    }
                }
                val aroundFuture = CompletableFutures.allAsList(chunksToLoad)

                this.future = this.future.thenCombineAsync(aroundFuture, PopulationTask.INSTANCE, chunkManager.executor)
                this.future.thenRun { POPULATION_RUNNING_UPDATER.compareAndSet(this, 1, 0) }
            }
        }

        fun finish() {
            populate()
            if ((this.chunk == null || !this.chunk!!.isFinished()) && FINISH_RUNNING_UPDATER.compareAndSet(
                    this,
                    0,
                    1
                )
            ) {
                val chunksToLoad = ArrayList<CompletableFuture<JukeboxChunk>>(8)
                for (z in this.z - 1..this.z + 1) {
                    for (x in this.x - 1..this.x + 1) {
                        if (x == this.x && z == this.z) continue
                        chunksToLoad.add(this.chunkManager.getChunkFuture(x, z,
                            generate = true,
                            populate = true,
                            finish = false
                        ))
                    }
                }
                val aroundFuture = CompletableFutures.allAsList(chunksToLoad)

                this.future = this.future.thenCombineAsync(aroundFuture, FinishingTask.INSTANCE, chunkManager.executor)
                this.future.thenRun { FINISH_RUNNING_UPDATER.compareAndSet(this, 1, 0) }
            }
        }

        private fun clear() {
            this.future = this.future.thenApply { loadedChunk ->
                GENERATION_RUNNING_UPDATER.set(this, 0)
                POPULATION_RUNNING_UPDATER.set(this, 0)
                FINISH_RUNNING_UPDATER.set(this, 0)
                loadedChunk
            }
        }
    }

}